design/matching_lineage.md

```diff
@@ -0,0 +1,114 @@
+        exists in the system, regardless of device, browser, or user
+      * A UUID is attached to the match to use to identify secondary data storage
+        (for DPs, for example)
+      * The UUID is strictly internal only and should not be visible in the API.
+        * This allows for atomic match deletion. Consider the following scenario:
+          * A match is created and the hash ID returned
+          * It's not used for a while and deletion of secondary data starts (which could take a
+            while)
+          * The same match is run again, and therefore the same hash ID returned
+          * A DP is accessed, which uses the match ID to look up secondary data and
+            either fails or returns incomplete data as it's mid deletion
+        * With a UUID for secondary data, the original match can be deleted atomically, and any
+          new match will get a new UUID, and so a DP will know it needs to recreate the data
+    * Matches are not expected to be searchable or otherwise accessible unless the user has
+      the match ID.
+  * Exact matches:
+    * The CS runs an `IN` AQL query on the collection attributes AC to find matching members of
+      the collection
+      * May need to be split into multiple queries if there are too many inputs
+    * This may return tens of thousands of matches per input G/A
+  * Rank matches:
+    * The user must also provide a rank level (e.g. phylum, domain, etc)
+    * The CS truncates the input matches' lineages to the provided rank
+    * The CS runs an `OR` query with a stanza for each truncated lineage with a
+      [starts with](https://github.com/arangodb/arangodb/issues/1796)
+      query to find all the members in the subtree
+      * An index on the lineage should still be able to support the starts with query
+        * **TODO**: test this assumption
+      * This might be pretty slow. We might have to do some experiements with data schemas and
+        queries.
+    * This may return tens of thousands of matches per input G/A
+  * Nearest matches:
+    * TBD. Don't have a clear understanding of this yet
+  * Matches
+    * Matches are not sharable and are ephemeral. They expire after X hours (X = 24?).
+      * X could be hardcoded, configured in the service configuration, or dynamically settable
+        via an API. TBD
+      * Matches must expire based on the *last usage* - e.g. if a match is retrieved, its lifetime
+        must be renewed.
+        * Perhaps up to some maximum lifetime, to avoid crafy people setting up cron jobs to
+          touch their matches.
+      * Will need some means of cleaning up the match and any supporting DP data produced (see
+        below), including ArangoDB records, cache service data, etc.
+  * Data visualization
+    * When accessing a DP (like `taxa_counts`) the user would provide the match ID
+    * The DP is responsible for any processing that is required to support the visualization,
```

Comment:
I think the wording here is a bit confusing.  The data product processing can and should do any pre-processing that can help to accelerate these queries, but the collections service will still be responsible for how to use any of that including what to cache.

Reply:
yeah, the DP I'm referring to here is a subsystem of the CS. I'll clarify that.

Reply:
Fixed. Also changed "connectors" to "matchers" as that seems to explain what they do better